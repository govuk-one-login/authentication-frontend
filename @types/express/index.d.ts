import type pino from "pino";
import type { Taxonomy } from "../utils/taxonomy.ts";

declare global {
  namespace Express {
    export interface Request {
      i18n?: {
        language?: string;
      };
      t: TFunction;
      csrfToken?: () => string;
      log: pino.Logger;
      generatedSessionId?: string;
    }

    export interface Locals {
      // Session IDs
      sessionId?: string;
      clientSessionId?: string;
      persistentSessionId?: string;
      clientId?: string;

      // Security
      scriptNonce?: string;
      csrfToken?: string;

      // Analytics
      ga4ContainerId?: string;
      isGa4Enabled?: string;
      analyticsCookieDomain?: string;
      analyticsProperties?: {
        httpStatusCode: number;
        contentId: string;
        taxonomy: Taxonomy;
      };

      // Language
      languageToggleEnabled?: boolean;
      language?: string;
      htmlLang?: string;

      // App
      strategicAppChannel?: boolean;
      webChannel?: boolean;
      genericAppChannel?: boolean;
      isApp?: boolean;

      // Feature flags
      supportPasskeyUsage?: boolean;
      supportPasskeyRegistration?: boolean;

      // Misc
      currentUrl?: URL;
      showTestBanner?: boolean;
      accountManagementUrl?: string;
      contactUsLinkUrl?: string;
    }
  }
}
