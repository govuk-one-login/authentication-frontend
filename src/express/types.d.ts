import 'express-serve-static-core';

declare global {
  namespace Express {
    interface Locals {
      sessionId: string;
      strategicAppChannel: boolean;
      webChannel: boolean;
      genericAppChannel: boolean;
      isApp: boolean
    }
    interface Response {
      locals: Locals;
    }
  }
}
