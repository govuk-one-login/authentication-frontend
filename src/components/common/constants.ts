import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
