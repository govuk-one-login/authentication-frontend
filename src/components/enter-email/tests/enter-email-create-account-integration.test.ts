import { describe } from "mocha";
import { expect, request, sinon } from "../../../../test/utils/test-utils.js";
import nock from "nock";
import * as cheerio from "cheerio";
import decache from "decache";
import {
  API_ENDPOINTS,
  HTTP_STATUS_CODES,
  PATH_NAMES,
} from "../../../app.constants.js";
import { ERROR_CODES, SecurityCodeErrorType } from "../../common/constants.js";
import { NextFunction, Request, Response } from "express";
import { getPermittedJourneyForPath } from "../../../../test/helpers/session-helper.js";
describe("Integration::enter email (create account)", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  before(async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = await import(
      "../../../middleware/session-middleware"
    );
    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (
        req: Request,
        res: Response,
        next: NextFunction
      ): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";

        req.session.user = {
          journey: getPermittedJourneyForPath(
            PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT
          ),
        };

        next();
      });

    app = await (await import("../../../app")).createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL;

    await request(
      app,
      (test) => test.get(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT),
      { expectAnalyticsPropertiesMatchSnapshot: false }
    ).then((res) => {
      const $ = cheerio.load(res.text);
      token = $("[name=_csrf]").val();
      cookies = res.headers["set-cookie"];
    });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    sinon.restore();
    app = undefined;
  });

  it("should return enter email page", async () => {
    await request(app, (test) =>
      test.get(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT).expect(200)
    );
  });

  it("should return error when csrf not present", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .send({
          email: "test@test.com",
        })
        .expect(403)
    );
  });

  it("should return validation error when email not entered", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#email-error").text()).to.contains(
            "Enter your email address"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when invalid email entered", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test.tÂµrn@example.com",
        })
        .expect(function (res) {
          const page = cheerio.load(res.text);
          expect(page("#email-error").text()).to.contains(
            "Enter an email address in the correct format, like name@example.com\n"
          );
        })
        .expect(400)
    );

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test.trnexample.com",
        })
        .expect(function (res) {
          const page = cheerio.load(res.text);
          expect(page("#email-error").text()).to.contains(
            "Enter an email address in the correct format, like name@example.com\n"
          );
        })
        .expect(400)
    );

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test.trn@examplecom",
        })
        .expect(function (res) {
          const page = cheerio.load(res.text);
          expect(page("#email-error").text()).to.contains(
            "Enter an email address in the correct format, like name@example.com\n"
          );
        })
        .expect(400)
    );
  });

  it("should redirect to /enter-password page when email address exists", async () => {
    nock(baseApi)
      .post(API_ENDPOINTS.USER_EXISTS)
      .once()
      .reply(HTTP_STATUS_CODES.OK, {
        email: "test@test.com",
        doesUserExist: true,
      });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test@test.com",
        })
        .expect("Location", PATH_NAMES.ENTER_PASSWORD_ACCOUNT_EXISTS)
        .expect(302)
    );
  });

  it("should redirect to /check-your-email when email address not found", async () => {
    nock(baseApi)
      .post(API_ENDPOINTS.USER_EXISTS)
      .once()
      .reply(200, {
        email: "test@test.com",
        userExists: false,
      })
      .post(API_ENDPOINTS.SEND_NOTIFICATION)
      .once()
      .reply(HTTP_STATUS_CODES.NO_CONTENT);

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test@test.com",
        })
        .expect("Location", PATH_NAMES.CHECK_YOUR_EMAIL)
        .expect(302)
    );
  });

  it("should return internal server error when /user-exists API call response is 500", async () => {
    nock(baseApi).post(API_ENDPOINTS.USER_EXISTS).once().reply(500, {
      message: "Internal Server error",
    });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test@test.com",
        })
        .expect(500)
    );
  });

  it("should redirect to /security-code-requested-too-many-times when exceeded OTP request limit", async () => {
    nock(baseApi)
      .post(API_ENDPOINTS.USER_EXISTS)
      .once()
      .reply(200, {
        email: "test@test.com",
        userExists: false,
      })
      .post(API_ENDPOINTS.SEND_NOTIFICATION)
      .once()
      .reply(400, { code: ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test@test.com",
        })
        .expect(
          "Location",
          `${PATH_NAMES.SECURITY_CODE_REQUEST_EXCEEDED}?actionType=${SecurityCodeErrorType.EmailBlocked}`
        )
        .expect(302)
    );
  });

  it("should redirect to /security-code-invalid-request when request OTP more than 5 times", async () => {
    nock(baseApi)
      .post(API_ENDPOINTS.USER_EXISTS)
      .once()
      .reply(200, {
        email: "test@test.com",
        userExists: false,
      })
      .post(API_ENDPOINTS.SEND_NOTIFICATION)
      .once()
      .reply(400, { code: ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          email: "test@test.com",
        })
        .expect((res) => {
          res.text.includes(
            "you asked to resend the security code too many codes"
          );
        })
        .expect(200)
    );
  });
});

describe("Integration::enter email (create account request)", () => {
  it(`should redirect to ${PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT} when ${PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT_REQUEST} is accessed`, async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = require("../../../middleware/session-middleware");
    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (
        req: Request,
        res: Response,
        next: NextFunction
      ): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";
        req.session.user = {
          journey: getPermittedJourneyForPath(PATH_NAMES.SIGN_IN_OR_CREATE),
        };
        next();
      });

    const app = await require("../../../app").createApp();

    await request(app, (test) =>
      test
        .get(PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT_REQUEST)
        .expect("Location", PATH_NAMES.ENTER_EMAIL_CREATE_ACCOUNT)
        .expect(302)
    );
  });
});
