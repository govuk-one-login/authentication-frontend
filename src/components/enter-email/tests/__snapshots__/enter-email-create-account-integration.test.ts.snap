// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration::enter email (create account) should redirect to /security-code-invalid-request when request OTP more than 5 times 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter email (create account) should return enter email page 1`] = `
Object {
  "contentId": "390f46f9-1f6b-44f2-8fd7-21a5385a7d3a",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter email (create account) should return error when csrf not present 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "",
  "taxonomyLevel2": "",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter email (create account) should return internal server error when /user-exists API call response is 500 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter email (create account) should return validation error when email not entered 1`] = `
Object {
  "contentId": "390f46f9-1f6b-44f2-8fd7-21a5385a7d3a",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter email (create account) should return validation error when invalid email entered 1`] = `
Object {
  "contentId": "390f46f9-1f6b-44f2-8fd7-21a5385a7d3a",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter email (create account) should return validation error when invalid email entered 2`] = `
Object {
  "contentId": "390f46f9-1f6b-44f2-8fd7-21a5385a7d3a",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter email (create account) should return validation error when invalid email entered 3`] = `
Object {
  "contentId": "390f46f9-1f6b-44f2-8fd7-21a5385a7d3a",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;
